src/apps/ENGINE/file_service.cpp

```diff
@@ -386,120 +386,156 @@ void INIFILE_T::WriteDouble(const char *section_name, const char *key_name, doub
 // fill buffer with key value, throw EXS exception object if failed or if section or key doesnt exist
 void INIFILE_T::ReadString(const char *section_name, const char *key_name, char *buffer, size_t buffer_size)
 {
+    SEARCH_DATA Search;
     ifs_PTR->ReadString(&Search, section_name, key_name, buffer, buffer_size);
 }
 
 // fill buffer with key value if section and key exist, otherwise fill with def_string and return false
 bool INIFILE_T::ReadString(const char *section_name, const char *key_name, char *buffer, size_t buffer_size,
                            const char *def_string)
 {
+    SEARCH_DATA Search;
     return ifs_PTR->ReadString(&Search, section_name, key_name, buffer, buffer_size, def_string);
 }
 
 // continue search from key founded in previous call this function or to function ReadString
 // fill buffer with key value if section and key exist, otherwise return false
-bool INIFILE_T::ReadStringNext(const char *section_name, const char *key_name, char *buffer, size_t buffer_size)
+std::size_t INIFILE_T::ForEachString(const char *section_name, const char *key_name,
+                                     std::function<bool(std::size_t, char *)> f)
 {
-    return ifs_PTR->ReadStringNext(&Search, section_name, key_name, buffer, buffer_size);
+    SEARCH_DATA s;
+
+    auto n = 0;
+    char buffer[2048];
+    while (true)
+    {
+        bool found;
+        if (n == 0)
+        {
+            found = ifs_PTR->ReadString(&s, section_name, key_name, buffer, 2048, "");
```

Comment:
it seems this can be moved out of the loop body. Same applies for the similar code in the other `ForEachString` overload.